### ISSUE TITLE
`WeightNorm` causes unexpected PyTrees inequality

### ISSUE THREAD

**Issue by yuanz271** [patrick-kidger/equinox#965]
The following assertion fails.
```
import equinox as eqx
import equinox.random as jrandom

layer = eqx.nn.Linear(2, 2, key=jrandom.key(0))
m1 = eqx.nn.WeightNorm(layer)
m2 = eqx.nn.WeightNorm(layer)
assert eqx.tree_equal(m1, m2)  # unequal
```
The reason of failure is that `WeightNorm._norm` is different (== operation returns False) for every `WeightNorm` instance.

This also causes the failure of using `tree_equal` to check deserialized models.

**Comment by johannahaffner** [comment]
This is expected. The values of the leaves are the same, but you have two new trees now, that makes two objects that JAX is treating separately. Static leaves (such as bound methods and their jaxprs) generally do not permit equality checks, and layer does have a bound method. If you partition the layers into the dynamic and static components, you will see that equality checks for the weights.

You can verify this:

```python
import equinox as eqx
import jax.random as jr


layer = eqx.nn.Linear(2, 2, key=jr.key(0))
m1 = eqx.nn.WeightNorm(layer)
m2 = eqx.nn.WeightNorm(layer)
d1, s1 = eqx.partition(m1, eqx.is_array)
d2, s2 = eqx.partition(m2, eqx.is_array)

assert eqx.tree_equal(d1, d2)  # Dynamic components (arrays) are equal
assert not eqx.tree_equal(s1, s2)  # Static components are not equal
```

**Comment by yuanz271** [comment]
@johannahaffner thank you.
> This is expected. The values of the leaves are the same, but you have two new trees now, that makes two objects that JAX is treating separately. Static leaves (such as bound methods and their jaxprs) generally do not permit equality checks, and layer does have a bound method. If you partition the layers into the dynamic and static components, you will see that equality checks for the weights.

However, what's the motivation to declare `_norm` in the inventory? The other instance (bound) methods are not compared.

**Comment by yuanz271** [comment]
Maybe I didn't ask clearly.
My point is that, if you define `_norm` as an instance method instead of a `Module` field, it will not be compared (like `__call__` of a `Module`). So, why this particular way?

**Comment by johannahaffner** [comment]
Ah! That is what you mean. This is consistent with how things are done elsewhere in the library. More complicated modules such as `MultiheadAttention` actually have several fields that are callable, e.g. here:

https://github.com/patrick-kidger/equinox/blob/8191b113df5d985720e86c0d6292bceb711cbe94/equinox/nn/_attention.py#L122

Comparing any callable for equality is just not what is expected in Python, and I don't think that there is a good reason to push things into methods in order to be able to do so.

**Comment by yuanz271** [comment]
That I understand because `query_proj` is a `Module` so that it handles well recursively. However, the `_norm` is internally supposed to be static and stateless by `WeightNorm`. Since `tree_equal` compares arrays by value, the semantics is more like to discriminate by its function rather than reference at least for a `Module`.

**Comment by johannahaffner** [comment]
Right. I can't speak for @patrick-kidger, but elsewhere in the Equinox system we do make frequent use of fields to specify some norm, e.g. [here](https://github.com/patrick-kidger/optimistix/blob/482f9a26d8e3b5f2cc3927ffc8fc9d9b554c24ad/optimistix/_solver/gauss_newton.py#L401). The reason being that a design like that allows to make the norm public as an optional input argument if requested, and users might then like to use a norm other than the Frobenius norm which `jnp.linalg.norm` defaults to. If this was implemented as an instance method, this would require a breaking change, if it is a field it just requires the addition of a keyword argument to `WeightNorm.__init__`, which can default to the current behaviour and would not be breaking.

**Comment by discobot** [comment]
Confirmed the cause is the per-instance `ft.partial` in `_norm`; one side effect not yet mentioned here is that it also makes `eqx.filter_jit` retrace once per `WeightNorm` instance, since each carries an identity-unequal leaf. Caching the partial on `axis` fixes both without changing the pytree structure. Fix in #1240.
